from PyQt6 import QtWidgets
from PyQt6.QtCore import QPoint, Qt
from PyQt6.QtCore import pyqtSignal
from PyQt6.QtGui import QPainter, QColor, QBrush, QPen, QFont
from PyQt6.QtWidgets import (QCalendarWidget, QVBoxLayout, QHBoxLayout,
                             QPushButton, QMessageBox,
                             QLabel, QFrame)

from simple_api_client import SimpleTheatreClient  # –î–æ–±–∞–≤–ª–µ–Ω –∏–º–ø–æ—Ä—Ç

from .base_page import BasePage


class CalendarWidget(QCalendarWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.lessons_data = {}
        self.setMouseTracking(True)

    def set_lessons_data(self, lessons_data):
        """–£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ—Ç –¥–∞–Ω–Ω—ã–µ –æ –∑–∞–Ω—è—Ç–∏—è—Ö"""
        self.lessons_data = lessons_data
        self.updateCells()

    def paintCell(self, painter, rect, date):
        """–ü–µ—Ä–µ–æ–ø—Ä–µ–¥–µ–ª—è–µ–º –æ—Ç—Ä–∏—Å–æ–≤–∫—É —è—á–µ–µ–∫ –∫–∞–ª–µ–Ω–¥–∞—Ä—è"""
        super().paintCell(painter, rect, date)

        date_str = date.toString("dd-MM-yyyy")
        if date_str in self.lessons_data:
            # –†–∏—Å—É–µ–º –∏–Ω–¥–∏–∫–∞—Ç–æ—Ä –∑–∞–Ω—è—Ç–∏–π
            painter.setRenderHint(QPainter.RenderHint.Antialiasing)

            # –ö—Ä–∞—Å–Ω—ã–π –∫—Ä—É–∂–æ–∫ —Å –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ–º –∑–∞–Ω—è—Ç–∏–π
            lessons_count = len(self.lessons_data[date_str])
            painter.setBrush(QBrush(QColor(255, 0, 0, 200)))
            painter.setPen(QPen(QColor(255, 255, 255)))

            circle_radius = 6
            circle_x = rect.right() - circle_radius - 3
            circle_y = rect.top() + circle_radius + 3

            # –†–∏—Å—É–µ–º –∫—Ä—É–∂–æ–∫
            painter.drawEllipse(
                circle_x - circle_radius,
                circle_y - circle_radius,
                circle_radius * 2,
                circle_radius * 2)

            # –†–∏—Å—É–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –∑–∞–Ω—è—Ç–∏–π
            painter.setPen(QPen(QColor(255, 255, 255)))
            painter.setFont(QFont("Arial", 6, QFont.Weight.Bold))
            painter.drawText(circle_x - 4, circle_y + 2, str(lessons_count))


class LessonPopup(QFrame):
    def __init__(self, parent=None):
        super().__init__(parent, Qt.WindowType.Popup)
        self.setFrameStyle(QFrame.Shape.Box)
        # self.setLineWidth(2)
        self.current_lesson_index = 0
        self.current_lessons = []
        self.current_date = ""
        self.init_ui()

    def init_ui(self):
        layout = QVBoxLayout()
        layout.setContentsMargins(10, 10, 10, 10)
        self.setStyleSheet("""
                    QFrame {
                        background-color: rgba(64, 20, 20, 1);
                        border-radius: 15px;
                        padding: 5px;
                        min-width: 365px;
                        border: none;
                    }
                    QFrame::window {
                        border-radius: 15px;
                    }
                    """)
        # –ó–∞–≥–æ–ª–æ–≤–æ–∫ —Å –Ω–∞–≤–∏–≥–∞—Ü–∏–µ–π
        header_layout = QHBoxLayout()

        self.nav_label = QLabel()
        self.nav_label.setStyleSheet(
            "font-family: 'Martian Mono Condensed'; font-size: 11px; color:#fdefc8;")
        header_layout.addWidget(self.nav_label)

        header_layout.addStretch()

        # –ö–Ω–æ–ø–∫–∞ –∑–∞–∫—Ä—ã—Ç–∏—è
        close_btn = QPushButton("√ó")
        close_btn.setFixedSize(20, 20)
        close_btn.setStyleSheet("""
            QPushButton {
                background-color: #dc3545;
                color: white;
                border: none;
                border-radius: 10px;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #c82333;
            }
        """)
        close_btn.clicked.connect(self.hide)
        header_layout.addWidget(close_btn)

        layout.addLayout(header_layout)

        # –ù–∞–≤–∏–≥–∞—Ü–∏—è –º–µ–∂–¥—É –∑–∞–Ω—è—Ç–∏—è–º–∏
        nav_layout = QHBoxLayout()

        self.prev_btn = QPushButton("‚óÄ –ü—Ä–µ–¥.")
        self.prev_btn.setStyleSheet("""
            QPushButton {
                font-family: "Martian Mono Condensed";
                font-size: 10px;
                color:#fdefc8;
                border-radius: 15px;
                background-color: rgba(113, 37, 37, 1);
                padding: 10px;
            }

            QPushButton:hover {
                    background-color: rgba(120, 41, 41, 1);
             }
            QPushButton:pressed {
                    background-color: rgba(85, 29, 29, 1);
                }
        """)
        self.prev_btn.clicked.connect(self.show_previous_lesson)
        nav_layout.addWidget(self.prev_btn)

        nav_layout.addStretch()

        self.next_btn = QPushButton("–°–ª–µ–¥. ‚ñ∂")
        self.next_btn.setStyleSheet("""
            QPushButton {
                font-family: "Martian Mono Condensed";
                font-size: 10px;
                color:#fdefc8;
                border-radius: 15px;
                background-color: rgba(113, 37, 37, 1);
                padding: 10px;
            }

            QPushButton:hover {
                    background-color: rgba(120, 41, 41, 1);
             }
            QPushButton:pressed {
                    background-color: rgba(85, 29, 29, 1);
                }
        """)
        self.next_btn.clicked.connect(self.show_next_lesson)
        nav_layout.addWidget(self.next_btn)

        layout.addLayout(nav_layout)

        # –ó–∞–≥–æ–ª–æ–≤–æ–∫ –∑–∞–Ω—è—Ç–∏—è
        self.title_label = QLabel()
        self.title_label.setStyleSheet(
            "font-family: 'Martian Mono Condensed'; font-size: 13px; color: #fdefc8; margin: 5px 0;")
        self.title_label.setWordWrap(True)
        layout.addWidget(self.title_label)

        # –î–µ—Ç–∞–ª–∏ –∑–∞–Ω—è—Ç–∏—è
        self.details_label = QLabel()
        self.details_label.setStyleSheet(
            "font-size: 11px; color: #fdefc8; font-family: 'Martian Mono Condensed';")
        self.details_label.setWordWrap(True)
        layout.addWidget(self.details_label)

        # –ö–Ω–æ–ø–∫–∞ —É–¥–∞–ª–µ–Ω–∏—è
        self.delete_btn = QPushButton("–£–¥–∞–ª–∏—Ç—å –∑–∞–Ω—è—Ç–∏–µ")
        self.delete_btn.setStyleSheet("""
            QPushButton {
                font-family: "Martian Mono Condensed";
                font-size: 10px;
                color:#fdefc8;
                border-radius: 15px;
                background-color: rgba(113, 37, 37, 1);
                padding: 10px;
            }

            QPushButton:hover {
                    background-color: rgba(120, 41, 41, 1);
             }
            QPushButton:pressed {
                    background-color: rgba(85, 29, 29, 1);
                }
        """)
        self.delete_btn.clicked.connect(self.delete_current_lesson)
        layout.addWidget(self.delete_btn)

        self.setLayout(layout)

    def show_lessons(self, lessons, date, pos):
        """–ü–æ–∫–∞–∑—ã–≤–∞–µ—Ç –∑–∞–Ω—è—Ç–∏—è –¥–ª—è –≤—ã–±—Ä–∞–Ω–Ω–æ–π –¥–∞—Ç—ã"""
        if not lessons:
            self.hide()
            return

        self.current_lessons = lessons
        self.current_date = date
        self.current_lesson_index = 0

        self.update_navigation()
        self.show_lesson_at_index()
        self.move(pos)
        self.show()
        self.adjustSize()

    def show_lesson_at_index(self):
        """–ü–æ–∫–∞–∑—ã–≤–∞–µ—Ç –∑–∞–Ω—è—Ç–∏–µ –ø–æ —Ç–µ–∫—É—â–µ–º—É –∏–Ω–¥–µ–∫—Å—É"""
        if not self.current_lessons:
            return

        lesson_data = self.current_lessons[self.current_lesson_index]
        lesson_id, title, description, time, location, date = lesson_data

        # –û–±–Ω–æ–≤–ª—è–µ–º –∑–∞–≥–æ–ª–æ–≤–æ–∫
        self.title_label.setText(f"üé≠ {title}")

        # –û–±–Ω–æ–≤–ª—è–µ–º –¥–µ—Ç–∞–ª–∏
        details = []
        if time:
            details.append(f"‚è∞ <b>–í—Ä–µ–º—è:</b> {time}")
        if location:
            details.append(f"üìç <b>–ú–µ—Å—Ç–æ:</b> {location}")
        if description:
            details.append(f"üìù <b>–û–ø–∏—Å–∞–Ω–∏–µ:</b> {description}")

        if not details:
            details.append("‚ÑπÔ∏è <i>–ù–µ—Ç –¥–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω–æ–π –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–∏</i>")

        self.details_label.setText("<br>".join(details))

        # –°–æ—Ö—Ä–∞–Ω—è–µ–º ID —Ç–µ–∫—É—â–µ–≥–æ –∑–∞–Ω—è—Ç–∏—è –¥–ª—è —É–¥–∞–ª–µ–Ω–∏—è
        self.current_lesson_id = lesson_id

    def update_navigation(self):
        """–û–±–Ω–æ–≤–ª—è–µ—Ç —Å–æ—Å—Ç–æ—è–Ω–∏–µ –∫–Ω–æ–ø–æ–∫ –Ω–∞–≤–∏–≥–∞—Ü–∏–∏"""
        total_lessons = len(self.current_lessons)

        # –û–±–Ω–æ–≤–ª—è–µ–º label –Ω–∞–≤–∏–≥–∞—Ü–∏–∏
        self.nav_label.setText(
            f"{self.current_lesson_index + 1}/{total_lessons}")

        # –û–±–Ω–æ–≤–ª—è–µ–º –∫–Ω–æ–ø–∫–∏
        self.prev_btn.setEnabled(self.current_lesson_index > 0)
        self.next_btn.setEnabled(self.current_lesson_index < total_lessons - 1)

    def show_previous_lesson(self):
        """–ü–æ–∫–∞–∑—ã–≤–∞–µ—Ç –ø—Ä–µ–¥—ã–¥—É—â–µ–µ –∑–∞–Ω—è—Ç–∏–µ"""
        if self.current_lesson_index > 0:
            self.current_lesson_index -= 1
            self.update_navigation()
            self.show_lesson_at_index()
            self.adjustSize()

    def show_next_lesson(self):
        """–ü–æ–∫–∞–∑—ã–≤–∞–µ—Ç —Å–ª–µ–¥—É—é—â–µ–µ –∑–∞–Ω—è—Ç–∏–µ"""
        if self.current_lesson_index < len(self.current_lessons) - 1:
            self.current_lesson_index += 1
            self.update_navigation()
            self.show_lesson_at_index()
            self.adjustSize()

    def delete_current_lesson(self):
        """–£–¥–∞–ª—è–µ—Ç —Ç–µ–∫—É—â–µ–µ –∑–∞–Ω—è—Ç–∏–µ"""
        reply = QMessageBox.question(
            self, "–ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–∏–µ —É–¥–∞–ª–µ–Ω–∏—è",
            "–í—ã —É–≤–µ—Ä–µ–Ω—ã, —á—Ç–æ —Ö–æ—Ç–∏—Ç–µ —É–¥–∞–ª–∏—Ç—å —ç—Ç–æ –∑–∞–Ω—è—Ç–∏–µ?",
            QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No
        )

        if reply == QMessageBox.StandardButton.Yes:
            if hasattr(self.parent(), 'delete_lesson_by_id'):
                self.parent().delete_lesson_by_id(self.current_lesson_id, self.current_date)


class ShedPage(BasePage):
    navigate_to = pyqtSignal(str)

    def __init__(self, parent=None):
        super().__init__("ui/shed_page.ui", parent)
        self.client = SimpleTheatreClient()  # –î–æ–±–∞–≤–ª–µ–Ω –∫–ª–∏–µ–Ω—Ç API
        self.current_user = ""
        self.init_db()

        self.setup_calendar()
        self.popup = LessonPopup(self)
        self.add_btn.clicked.connect(self.add_lesson)
        self.load_lessons_data()
        is_organizer = self.current_user == "organizer"
        if not is_organizer:
            self.form_container.setVisible(is_organizer)
            self.calendarWidget.move(150, 0)

    def init_db(self):
        """–ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö –¥–ª—è –∑–∞–Ω—è—Ç–∏–π —á–µ—Ä–µ–∑ API"""
        user_data = self.client.get_current_user()
        if user_data:
            self.current_user = user_data['role']
    def setup_calendar(self):
        """–ù–∞—Å—Ç—Ä–æ–π–∫–∞ –∫–∞–ª–µ–Ω–¥–∞—Ä—è"""

        # –ó–∞–º–µ–Ω—è–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π –∫–∞–ª–µ–Ω–¥–∞—Ä—å –Ω–∞ –Ω–∞—à –∫–∞—Å—Ç–æ–º–Ω—ã–π
        old_calendar = self.calendarWidget
        self.calendarWidget = CalendarWidget(self)
        self.calendarWidget.setGeometry(old_calendar.geometry())
        self.calendarWidget.setStyleSheet(old_calendar.styleSheet())
        self.calendarWidget.setGridVisible(False)
        self.calendarWidget.setNavigationBarVisible(True)
        self.calendarWidget.setDateEditEnabled(True)
        self.calendarWidget.setVerticalHeaderFormat(
            QtWidgets.QCalendarWidget.VerticalHeaderFormat.NoVerticalHeader)

        # –ü–æ–¥–∫–ª—é—á–∞–µ–º –æ–±—Ä–∞–±–æ—Ç—á–∏–∫–∏
        self.calendarWidget.clicked.connect(self.on_date_clicked)

        # –£–¥–∞–ª—è–µ–º —Å—Ç–∞—Ä—ã–π –∫–∞–ª–µ–Ω–¥–∞—Ä—å
        old_calendar.deleteLater()

    def load_lessons_data(self):
        """–ó–∞–≥—Ä—É–∂–∞–µ—Ç –¥–∞–Ω–Ω—ã–µ –æ –∑–∞–Ω—è—Ç–∏—è—Ö –¥–ª—è –∫–∞–ª–µ–Ω–¥–∞—Ä—è —á–µ—Ä–µ–∑ API"""
        try:
            lessons = self.client.get_lessons()

            # –ì—Ä—É–ø–ø–∏—Ä—É–µ–º –∑–∞–Ω—è—Ç–∏—è –ø–æ –¥–∞—Ç–∞–º
            lessons_by_date = {}
            for lesson in lessons:
                date = lesson['date']
                if date not in lessons_by_date:
                    lessons_by_date[date] = []
                lessons_by_date[date].append((
                    lesson['id'],
                    lesson['title'],
                    lesson['description'],
                    lesson['time'],
                    lesson['location'],
                    lesson['date']
                ))

            self.calendarWidget.set_lessons_data(lessons_by_date)

        except Exception as e:
            print(f"–û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ –¥–∞–Ω–Ω—ã—Ö: {e}")

    def on_date_clicked(self, date):
        """–û–±—Ä–∞–±–æ—Ç–∫–∞ –∫–ª–∏–∫–∞ –ø–æ –¥–∞—Ç–µ –≤ –∫–∞–ª–µ–Ω–¥–∞—Ä–µ"""
        date_str = date.toString("dd-MM-yyyy")

        # –ü–æ–ª—É—á–∞–µ–º –∑–∞–Ω—è—Ç–∏—è –Ω–∞ —ç—Ç—É –¥–∞—Ç—É —á–µ—Ä–µ–∑ API
        lessons = self.client.get_lessons()
        date_lessons = [lesson for lesson in lessons if lesson['date'] == date_str]

        if date_lessons:
            # –ö–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–µ–º –≤ –Ω—É–∂–Ω—ã–π —Ñ–æ—Ä–º–∞—Ç –¥–ª—è –ø–æ–ø–∞–ø–∞
            formatted_lessons = []
            for lesson in date_lessons:
                formatted_lessons.append((
                    lesson['id'],
                    lesson['title'],
                    lesson['description'],
                    lesson['time'],
                    lesson['location'],
                    lesson['date']
                ))

            # –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –ø–æ–ø–∞–ø —Å –∑–∞–Ω—è—Ç–∏—è–º–∏ —Ä—è–¥–æ–º —Å –∫–∞–ª–µ–Ω–¥–∞—Ä–µ–º
            calendar_pos = self.calendarWidget.mapToGlobal(QPoint(0, 0))
            popup_pos = QPoint(calendar_pos.x(), calendar_pos.y() + 400)

            self.popup.show_lessons(formatted_lessons, date_str, popup_pos)

    def delete_lesson_by_id(self, lesson_id, date):
        """–£–¥–∞–ª—è–µ—Ç –∑–∞–Ω—è—Ç–∏–µ –ø–æ ID —á–µ—Ä–µ–∑ API"""
        try:
            success = self.client.delete_lesson(lesson_id)
            if success:
                # –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –¥–∞–Ω–Ω—ã—Ö
                self.load_lessons_data()
                self.popup.hide()
                QMessageBox.information(self, "–£—Å–ø–µ—Ö", "–ó–∞–Ω—è—Ç–∏–µ —É–¥–∞–ª–µ–Ω–æ")
            else:
                QMessageBox.warning(self, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å —É–¥–∞–ª–∏—Ç—å –∑–∞–Ω—è—Ç–∏–µ")

        except Exception as e:
            QMessageBox.critical(self, "–û—à–∏–±–∫–∞", f"–û—à–∏–±–∫–∞ —É–¥–∞–ª–µ–Ω–∏—è: {e}")

    def add_lesson(self):
        """–î–æ–±–∞–≤–ª–µ–Ω–∏–µ –Ω–æ–≤–æ–≥–æ –∑–∞–Ω—è—Ç–∏—è —á–µ—Ä–µ–∑ API"""
        title = self.title_input.toPlainText().strip()
        time = self.time_input.text().strip()
        selected_date = self.calendarWidget.selectedDate().toString("dd-MM-yyyy")

        if not title:
            QMessageBox.warning(self, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥–∏—Ç–µ –Ω–∞–∑–≤–∞–Ω–∏–µ –∑–∞–Ω—è—Ç–∏—è")
            return

        try:
            success = self.client.create_lesson(title, selected_date, time, "", "")
            if success:
                # –û—á–∏—Å—Ç–∫–∞ —Ñ–æ—Ä–º—ã
                self.title_input.clear()
                self.time_input.clear()

                # –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –¥–∞–Ω–Ω—ã—Ö
                self.load_lessons_data()

                QMessageBox.information(
                    self, "–£—Å–ø–µ—Ö", f"–ó–∞–Ω—è—Ç–∏–µ '{title}' –¥–æ–±–∞–≤–ª–µ–Ω–æ –Ω–∞ {selected_date}!")
            else:
                QMessageBox.warning(self, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å –∑–∞–Ω—è—Ç–∏–µ")

        except Exception as e:
            QMessageBox.critical(self, "–û—à–∏–±–∫–∞", f"–û—à–∏–±–∫–∞: {e}")
